### ISSUE TITLE
Bug in bridges with variables-constrained-on-creation

### ISSUE THREAD

**Issue by chriscoey** [jump-dev/MathOptInterface.jl#1698]
I'm on latest JuMP and MOI. This occurred running MOIPajarito's soc2 test here https://github.com/chriscoey/MOIPajarito.jl/blob/336edcf24cb3ee18dddbfaa955e9071a562848c9/test/JuMP_tests.jl#L87.
```
_soc2: Error During Test at /home/coey/.julia/dev/MOIPajarito/test/JuMP_tests.jl:29
  Got exception outside of a @test
  In `MathOptInterface.ScalarAffineFunction{Float64}`-in-`MathOptInterface.EqualTo{Float64}` constraint: Constant 2.5 of the function is not zero. The function constant should be moved to the set. You can use `MOI.Utilities.normalize_and_add_constraint` which does this automatically.
  Stacktrace:
    [1] throw_if_scalar_and_constant_not_zero
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/constraints.jl:95 [inlined]
    [2] bridged_constraint_function
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:1925 [inlined]
    [3] add_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.EqualTo{Float64})
      @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:1533
    [4] bridge_constraint(#unused#::Type{MathOptInterface.Bridges.Constraint.ScalarSlackBridge{Float64, MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.Integer}}, model::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.Integer)
      @ MathOptInterface.Bridges.Constraint ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/Constraint/slack.jl:182
    [5] add_bridged_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, BridgeType::Type, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.Integer)
      @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:1469
    [6] add_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.Integer)
      @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:1544
    [7] bridge_constraint(#unused#::Type{MathOptInterface.Bridges.Constraint.ScalarFunctionizeBridge{Float64, MathOptInterface.Integer}}, model::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)
      @ MathOptInterface.Bridges.Constraint ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/Constraint/functionize.jl:20
    [8] add_bridged_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, BridgeType::Type, f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)
      @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:1469
    [9] add_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)
      @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:1521
   [10] _copy_constraints(dest::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, src::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, index_map::MathOptInterface.Utilities.IndexMap, cis_src::Vector{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer}})
      @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/copy.jl:248
   [11] _pass_constraints(dest::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, src::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, index_map::MathOptInterface.Utilities.IndexMap, variable_constraints_not_added::Vector{Any})
      @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/copy.jl:303
   [12] default_copy_to(dest::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, src::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}})
      @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/copy.jl:442
   [13] #copy_to#7
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:421 [inlined]
   [14] copy_to
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:421 [inlined]
   [15] optimize!
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/MathOptInterface.jl:80 [inlined]
   [16] optimize!(m::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}})
      @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/cachingoptimizer.jl:285
   [17] optimize!(model::JuMP.Model, optimizer_factory::Nothing; bridge_constraints::Bool, ignore_optimize_hook::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ JuMP ~/.julia/packages/JuMP/2IF9U/src/optimizer_interface.jl:195
   [18] optimize! (repeats 2 times)
      @ ~/.julia/packages/JuMP/2IF9U/src/optimizer_interface.jl:167 [inlined]
   [19] _soc2(opt::MathOptInterface.OptimizerWithAttributes)
      @ Main.TestJuMP ~/.julia/dev/MOIPajarito/test/JuMP_tests.jl:87
```
Not sure how to get an MWE. I don't understand what is going on, but it seems like it is trying to bridge a VariableIndex in Integer to a ScalarAffineFunction in Integer, which doesn't make sense.

**Comment by odow** [comment]
Why is it trying to bridge `f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)`? Did you declare `supports_constraint` appropriately?

**Comment by chriscoey** [comment]
AFAICT yes, but IDK

**Comment by chriscoey** [comment]
it occurs with 
```
@variable(m, z <= 2.5, Int)
```
but not with
```
@variable(m, z, Int)
@constraint(m, z <= 2.5)
```

**Comment by chriscoey** [comment]
but changing that, I now get another error here: https://github.com/chriscoey/MOIPajarito.jl/blob/d2a7a25e452d536eb85c223b377db4a83954c1d5/test/JuMP_tests.jl#L100

as follows:
```
_soc2: Error During Test at /Users/coey/.julia/dev/MOIPajarito/test/JuMP_tests.jl:29
  Got exception outside of a @test
  KeyError: key MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer}(1) not found
  Stacktrace:
    [1] (::MathOptInterface.Utilities.DoubleDicts.var"#1#2"{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer}})()
      @ MathOptInterface.Utilities.DoubleDicts ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/DoubleDicts.jl:181
    [2] get
      @ ./dict.jl:547 [inlined]
    [3] getindex
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/DoubleDicts.jl:180 [inlined]
    [4] getindex
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/DoubleDicts.jl:173 [inlined]
    [5] getindex
      @ ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/copy/index_map.jl:61 [inlined]
    [6] delete(m::MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, index::MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer})
      @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/cachingoptimizer.jl:659
    [7] delete(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, ci::MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer})
      @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/hcTsY/src/Bridges/bridge_optimizer.jl:588
    [8] delete(m::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}, index::MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer})
      @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/hcTsY/src/Utilities/cachingoptimizer.jl:662
    [9] delete(model::JuMP.Model, con_ref::JuMP.ConstraintRef{JuMP.Model, MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer}, JuMP.ScalarShape})
      @ JuMP ~/.julia/packages/JuMP/2IF9U/src/constraints.jl:292
   [10] unset_integer
      @ ~/.julia/packages/JuMP/2IF9U/src/variables.jl:862 [inlined]
   [11] _soc2(opt::MathOptInterface.OptimizerWithAttributes)
      @ Main.TestJuMP ~/.julia/dev/MOIPajarito/test/JuMP_tests.jl:100
```

**Comment by odow** [comment]
I'll need to take a closer look at MOIPajarito. It shouldn't try to bridge VariableIndex in Integer.

But this doesn't seem right: https://github.com/chriscoey/MOIPajarito.jl/blob/d2a7a25e452d536eb85c223b377db4a83954c1d5/src/MOI_wrapper.jl#L61-L63
because you always query the integer variables
https://github.com/chriscoey/MOIPajarito.jl/blob/d2a7a25e452d536eb85c223b377db4a83954c1d5/src/MOI_wrapper.jl#L100-L101

**Comment by chriscoey** [comment]
What's wrong with doing that? MOIPajarito hasn't been failing on continuous problems

**Comment by odow** [comment]
Is this still an issue? I can't reproduce it on MOIPajarito.

**Comment by chriscoey** [comment]
I commented out the failing tests so you won't see it by running the current tests. There are two different errors referred to above. Did you run the code that I linked to for both errors?

**Comment by odow** [comment]
These are not the same. The second adds a linear constraint, the first a variable bound.
```Julia
@variable(m, z <= 2.5, Int)
```

```julia
@variable(m, z, Int)
@constraint(m, z <= 2.5)
```

I have 3 PRs open to MOIPajarito that fix some things

**Comment by odow** [comment]
Okay, I'll take a look at this:
```Julia
julia> using JuMP, GLPK, ECOS, MOIPajarito, Test

julia> m = Model(
           optimizer_with_attributes(
               MOIPajarito.Optimizer,
               "oa_solver" => MOI.OptimizerWithAttributes(GLPK.Optimizer),
               "conic_solver" => MOI.OptimizerWithAttributes(ECOS.Optimizer),
           )
       )
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Pajarito

julia> @variable(m, z <= 2.5, Int)
z

julia> optimize!(m)
ERROR: In `MathOptInterface.ScalarAffineFunction{Float64}`-in-`MathOptInterface.EqualTo{Float64}` constraint: Constant 2.5 of the function is not zero. The function constant should be moved to the set. You can use `MOI.Utilities.normalize_and_add_constraint` which does this automatically.
Stacktrace:
  [1] throw_if_scalar_and_constant_not_zero
    @ ~/.julia/packages/MathOptInterface/eoIu0/src/constraints.jl:95 [inlined]
  [2] bridged_constraint_function
    @ ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:1925 [inlined]
  [3] add_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.EqualTo{Float64})
    @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:1533
  [4] bridge_constraint(#unused#::Type{MathOptInterface.Bridges.Constraint.ScalarSlackBridge{Float64, MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.Integer}}, model::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.Integer)
    @ MathOptInterface.Bridges.Constraint ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/Constraint/slack.jl:182
  [5] add_bridged_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, BridgeType::Type, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.Integer)
    @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:1469
  [6] add_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.ScalarAffineFunction{Float64}, s::MathOptInterface.Integer)
    @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:1544
  [7] bridge_constraint(#unused#::Type{MathOptInterface.Bridges.Constraint.ScalarFunctionizeBridge{Float64, MathOptInterface.Integer}}, model::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)
    @ MathOptInterface.Bridges.Constraint ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/Constraint/functionize.jl:20
  [8] add_bridged_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, BridgeType::Type, f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)
    @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:1469
  [9] add_constraint(b::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, f::MathOptInterface.VariableIndex, s::MathOptInterface.Integer)
    @ MathOptInterface.Bridges ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:1521
 [10] _copy_constraints(dest::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, src::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, index_map::MathOptInterface.Utilities.IndexMap, cis_src::Vector{MathOptInterface.ConstraintIndex{MathOptInterface.VariableIndex, MathOptInterface.Integer}})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/eoIu0/src/Utilities/copy.jl:248
 [11] _pass_constraints(dest::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, src::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, index_map::MathOptInterface.Utilities.IndexMap, variable_constraints_not_added::Vector{Any})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/eoIu0/src/Utilities/copy.jl:303
 [12] default_copy_to(dest::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, src::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/eoIu0/src/Utilities/copy.jl:442
 [13] #copy_to#7
    @ ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:421 [inlined]
 [14] copy_to
    @ ~/.julia/packages/MathOptInterface/eoIu0/src/Bridges/bridge_optimizer.jl:421 [inlined]
 [15] optimize!
    @ ~/.julia/packages/MathOptInterface/eoIu0/src/MathOptInterface.jl:80 [inlined]
 [16] optimize!(m::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{MOIPajarito.Optimizer, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}, MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}})
    @ MathOptInterface.Utilities ~/.julia/packages/MathOptInterface/eoIu0/src/Utilities/cachingoptimizer.jl:285
 [17] optimize!(model::Model, optimizer_factory::Nothing; bridge_constraints::Bool, ignore_optimize_hook::Bool, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JuMP ~/.julia/packages/JuMP/2IF9U/src/optimizer_interface.jl:195
 [18] optimize! (repeats 2 times)
    @ ~/.julia/packages/JuMP/2IF9U/src/optimizer_interface.jl:167 [inlined]
 [19] top-level scope
    @ REPL[148]:1
```

**Comment by chriscoey** [comment]
Thanks, merged the PRs and added you as a collaborator

**Comment by odow** [comment]
Okay @blegat this is an interesting edge-case. Pajarito doesn't support variable bounds, so `x in GreaterThan(1.0)` gets bridged to `y in Nonnegatives`, where `y = x - 1`. But then `x in Integer` gets replaced as `y + 1 in Integer`.

In this case I want to opt-out of constrained on creation because `x free` and `x - 1 in Nonnegatives` is the better formulation.

**Comment by odow** [comment]
If you copy the constraints in the order `(Integer, MOI.GreaterThan{Float64})` then things work well. I need to keep digging.

**Comment by odow** [comment]
See #1699 for some changes in the set order. 

A related problem is that because we know we'll have to bridge `GreaterThan` at some point, `add_constrained_variable(model, MOI.GreaterThan(1.0))` uses a variable bridge instead of adding the free variable and then adding (then bridging) the bound post-facto. It seems like we should really say, "can I bridge this as a constraint" and if so, prefer that over adding it that way.

**Comment by odow** [comment]
Closed by #1699
